File: other/Yandex - Тренировки по алгоритмам 6.0/3/J.py
import random
import itertools
import sys

n, W = map(int, sys.stdin.readline().split())
h_arr = list(map(int, sys.stdin.readline().split()))
w_arr = list(map(int, sys.stdin.readline().split()))

def get_res():
    if max(w_arr) >= W or n == 1:
        return 0
    combined_arr = sorted(list(zip(h_arr, w_arr)))

    diffs = [0] * (n - 1)
    for i in range(1, n):
        diffs[i - 1] = combined_arr[i][0] - combined_arr[i - 1][0]
    
    idx_of_closest_larger_element_to_the_right = [None] * len(diffs)
    stack = []
    for i in range(len(diffs) - 1, -1, -1):
        while stack and diffs[stack[-1]] <= diffs[i]:
            stack.pop()
        idx_of_closest_larger_element_to_the_right[i] = stack[-1] if stack else None
        stack.append(i)
    idx_of_closest_larger_element_to_the_left = [None] * len(diffs)
    stack = []
    for i in range(len(diffs)):
        while stack and diffs[stack[-1]] <= diffs[i]:
            stack.pop()
        idx_of_closest_larger_element_to_the_left[i] = stack[-1] if stack else None
        stack.append(i) 

    prefix_sums = [0] * n
    cur_sum = 0
    for i in range(n):
        cur_sum += combined_arr[i][1]
        prefix_sums[i] = cur_sum
    # print(combined_arr)
    # print(prefix_sums)
    left, right = 0, 1
    res = float('inf')
    while right < n:
        max_diff_allowed = combined_arr[right][0] - combined_arr[left][0]
        this_diff_idx = left 
        first_larger_to_the_right_idx = idx_of_closest_larger_element_to_the_right[this_diff_idx]
        first_larger_to_the_left_idx = idx_of_closest_larger_element_to_the_left[this_diff_idx]
        last_el_idx = n - 1
        first_el_idx = 0
        if first_larger_to_the_right_idx is not None:
            last_el_idx = first_larger_to_the_right_idx
        if first_larger_to_the_left_idx is not None:
            first_el_idx = first_larger_to_the_left_idx + 1
        
        # print(left, right, first_el_idx, last_el_idx)
        width = prefix_sums[last_el_idx] - prefix_sums[first_el_idx - 1] if first_el_idx != 0 else prefix_sums[last_el_idx]
        if width >= W:
            res = min(res, max_diff_allowed)
        left += 1
        right += 1
    return res

print(get_res())

def my_func(n, W, h_arr, w_arr):

    def get_res():
        if max(w_arr) >= W or n == 1:
            return 0
        combined_arr = sorted(list(zip(h_arr, w_arr)))

        diffs = [0] * (n - 1)
        for i in range(1, n):
            diffs[i - 1] = combined_arr[i][0] - combined_arr[i - 1][0]

        idx_of_closest_larger_element_to_the_right = [None] * len(diffs)
        stack = []
        for i in range(len(diffs) - 1, -1, -1):
            while stack and diffs[stack[-1]] <= diffs[i]:
                stack.pop()
            idx_of_closest_larger_element_to_the_right[i] = stack[-1] if stack else None
            stack.append(i)
        idx_of_closest_larger_element_to_the_left = [None] * len(diffs)
        stack = []
        for i in range(len(diffs)):
            while stack and diffs[stack[-1]] <= diffs[i]:
                stack.pop()
            idx_of_closest_larger_element_to_the_left[i] = stack[-1] if stack else None
            stack.append(i) 

        prefix_sums = [0] * n
        cur_sum = 0
        for i in range(n):
            cur_sum += combined_arr[i][1]
            prefix_sums[i] = cur_sum
        # print(combined_arr)
        # print(prefix_sums)
        left, right = 0, 1
        res = float('inf')
        while right < n:
            max_diff_allowed = combined_arr[right][0] - combined_arr[left][0]
            this_diff_idx = left 
            first_larger_to_the_right_idx = idx_of_closest_larger_element_to_the_right[this_diff_idx]
            first_larger_to_the_left_idx = idx_of_closest_larger_element_to_the_left[this_diff_idx]
            last_el_idx = n - 1
            first_el_idx = 0
            if first_larger_to_the_right_idx is not None:
                last_el_idx = first_larger_to_the_right_idx
            if first_larger_to_the_left_idx is not None:
                first_el_idx = first_larger_to_the_left_idx + 1
            
            # print(left, right, first_el_idx, last_el_idx)
            width = prefix_sums[last_el_idx] - prefix_sums[first_el_idx - 1] if first_el_idx != 0 else prefix_sums[last_el_idx]
            if width >= W:
                res = min(res, max_diff_allowed)
            left += 1
            right += 1
        return res

    return(get_res())

def brute_force(n, W, h_arr, w_arr):
    res = float('inf')
    def generate_permutations(max_number):
        for r in range(1, max_number + 1):
            for perm in itertools.permutations(range(max_number), r):
                yield tuple(perm)
    
    for perm in generate_permutations(n):
        cur_w = 0
        cur_h = 0
        for idx in perm:
            cur_w += w_arr[idx]

        if len(perm) > 1:
            left, right = 0, 1
            while right < len(perm):
                cur_h = max(cur_h, abs(h_arr[perm[right]] - h_arr[perm[left]]))
                left += 1
                right += 1
        if cur_w >= W:
            res = min(res, cur_h)
        last_perm = perm
        
    return res, last_perm

        

def check():
    tests_conducted = 0
    wrongs_tests = 0
    for _ in range(100):
        n = random.randint(1, 5)
        h_arr = [random.randint(1, 100) for _ in range(n)]
        w_arr = [random.randint(1, 100) for _ in range(n)]
        W = random.randint(1, 100)
        if sum(w_arr) < W:
            continue
        tests_conducted += 1
        my_res = my_func(n, W, h_arr, w_arr)
        brute_force_res, brute_force_perm = brute_force(n, W, h_arr, w_arr)
        if my_res != brute_force_res:
            wrongs_tests += 1
            print(n, W, h_arr, w_arr)
            print(my_res, brute_force_res)
            print(brute_force_perm)
            break
    print(f'Tests conducted: {tests_conducted}')
    print(f'Wrong tests: {wrongs_tests}')

# check()

File: other/Yandex - Тренировки по алгоритмам 6.0/3/J.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>
#include <limits>

using namespace std;

int main() {
    int n;
    long long W;
    cin >> n >> W;
    
    vector<long long> h_arr(n), w_arr(n);
    for(int i = 0; i < n; i++) cin >> h_arr[i];
    for(int i = 0; i < n; i++) cin >> w_arr[i];
    
    if(*max_element(w_arr.begin(), w_arr.end()) >= W || n == 1) {
        cout << 0 << endl;
        return 0;
    }

    vector<pair<long long, long long>> combined_arr(n);
    for(int i = 0; i < n; i++) {
        combined_arr[i] = {h_arr[i], w_arr[i]};
    }
    sort(combined_arr.begin(), combined_arr.end());

    vector<long long> diffs(n - 1);
    for(int i = 1; i < n; i++) {
        diffs[i - 1] = combined_arr[i].first - combined_arr[i - 1].first;
    }

    vector<int> idx_of_closest_larger_element_to_the_right(n - 1, -1);
    stack<int> st;
    for(int i = n - 2; i >= 0; i--) {
        while(!st.empty() && diffs[st.top()] <= diffs[i]) {
            st.pop();
        }
        idx_of_closest_larger_element_to_the_right[i] = st.empty() ? -1 : st.top();
        st.push(i);
    }

    vector<int> idx_of_closest_larger_element_to_the_left(n - 1, -1);
    while(!st.empty()) st.pop();
    for(int i = 0; i < n - 1; i++) {
        while(!st.empty() && diffs[st.top()] <= diffs[i]) {
            st.pop();
        }
        idx_of_closest_larger_element_to_the_left[i] = st.empty() ? -1 : st.top();
        st.push(i);
    }

    vector<long long> prefix_sums(n);
    long long cur_sum = 0;
    for(int i = 0; i < n; i++) {
        cur_sum += combined_arr[i].second;
        prefix_sums[i] = cur_sum;
    }

    int left = 0, right = 1;
    long long res = numeric_limits<long long>::max();
    while(right < n) {
        long long max_diff_allowed = combined_arr[right].first - combined_arr[left].first;
        int this_diff_idx = left;
        int first_larger_to_the_right_idx = idx_of_closest_larger_element_to_the_right[this_diff_idx];
        int first_larger_to_the_left_idx = idx_of_closest_larger_element_to_the_left[this_diff_idx];
        int last_el_idx = n - 1;
        int first_el_idx = 0;

        if(first_larger_to_the_right_idx != -1) {
            last_el_idx = first_larger_to_the_right_idx;
        }
        if(first_larger_to_the_left_idx != -1) {
            first_el_idx = first_larger_to_the_left_idx + 1;
        }

        long long width = first_el_idx == 0 ? prefix_sums[last_el_idx] : 
                         prefix_sums[last_el_idx] - prefix_sums[first_el_idx - 1];
        
        if(width >= W) {
            res = min(res, max_diff_allowed);
        }
        left++;
        right++;
    }

    cout << res << endl;
    return 0;
}